src/main/credentialServer/httpCredentialHelperServer.ts

```diff
@@ -0,0 +1,316 @@
+import http from 'http';
+import path from 'path';
+import { URL } from 'url';
+import childProcess from 'child_process';
+import util from 'util';
+
+import Logging from '@/utils/logging';
+import paths from '@/utils/paths';
+import * as serverHelper from '@/main/serverHelper';
+import { findHomeDir } from '@/config/findHomeDir';
+import { spawnFile } from '@/utils/childProcess';
+
+export type ServerState = {
+  user: string;
+  password: string;
+  port: number;
+  pid: number;
+}
+
+const console = Logging.server;
+const SERVER_PORT = 6109;
+const SERVER_USERNAME = 'user';
+const SERVER_FILE_BASENAME = 'credential-server.json';
+const MAX_REQUEST_BODY_LENGTH = 2048;
+const isWindows = os.platform().startsWith('win');
+
+type dispatchFunctionType = (helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse) => Promise<void>;
+
+type checkFunctionOutputType = (stdout: string) => boolean;
+
+function requireNoOutput(stdout: string): boolean {
+  return !stdout;
+}
+
+function requireJSONOutput(stdout: string): boolean {
+  try {
+    JSON.parse(stdout);
+
+    return true;
+  } catch {}
+
+  return false;
+}
+
+export function getServerCredentialsPath(): string {
+  return path.join(paths.appHome, SERVER_FILE_BASENAME);
+}
+
+export class HttpCredentialHelperServer {
+  protected servers = isWindows ? [http.createServer(), http.createServer()] : [http.createServer()];
+  protected password = serverHelper.randomStr();
+  protected stateInfo: ServerState = {
+    user:     SERVER_USERNAME,
+    password: this.password,
+    port:     SERVER_PORT,
+    pid:      process.pid,
+  };
+
+  protected dispatchTable: Record<string, Record<string, dispatchFunctionType>> = {
+    POST: {
+      get:   this.get,
+      store: this.store,
+      erase: this.erase,
+      list:  this.list,
+    },
+  };
+
+  async init() {
+    const statePath = getServerCredentialsPath();
+
+    await fs.promises.writeFile(statePath,
+      JSON.stringify(this.stateInfo, undefined, 2),
+      { mode: 0o600 });
+    for (const server of this.servers) {
+      server.on('request', this.handleRequest.bind(this));
+      server.on('error', (err) => {
+        console.error(`Error writing out ${ statePath }`, err);
+      });
+    }
+    this.servers[0].listen(SERVER_PORT, '127.0.0.1');
+    if (isWindows) {
+      const addr = await getWSLPort();
+
+      if (addr) {
+        this.servers[1].listen(SERVER_PORT, addr);
+      } else {
+        this.servers.splice(1);
+      }
+    }
+    console.log('Credentials server is now ready.');
+  }
+
+  protected async handleRequest(request: http.IncomingMessage, response: http.ServerResponse) {
+    try {
+      if (!serverHelper.basicAuth(SERVER_USERNAME, this.password, request.headers.authorization ?? '')) {
+        response.writeHead(401, { 'Content-Type': 'text/plain' });
+
+        return;
+      }
+      const helperName = `docker-credential-${ await this.getCredentialHelperName() }`;
```

Comment:
It's executed in a top-level try/catch/finally block. I didn't see a need to specifically catch a failure here - it will show up in the 500-level error message.